modules/evs/fsw/src/cfe_evs_utils.c

```diff
@@ -288,6 +289,92 @@ bool EVS_IsFiltered(EVS_AppData_t *AppDataPtr, uint16 EventID, uint16 EventType)
+ *
+ * Function: EVS_SendEvent
+ *
+ * Application-scope internal function
+ * See description in header file for argument/return detail
+ *
+ *-----------------------------------------------------------------*/
+bool EVS_CheckAndIncrementSquelchTokens(EVS_AppData_t *AppDataPtr)
+{
+    bool               NotSquelched = true;
+    CFE_TIME_SysTime_t CurrentTime;
+    CFE_TIME_SysTime_t DeltaTime;
+    uint16_t           ReturnCount = 0;
+    char               AppName[OS_MAX_API_NAME];
+
+    if (CFE_PLATFORM_EVS_MAX_APP_EVENT_BURST != 0)
+    {
+        /*
+	 * We use a timer here since configurations are not guaranteed to send EVS HK wakeups at 1Hz
+	 */
+        CurrentTime = CFE_TIME_GetTime();
+        DeltaTime   = CFE_TIME_Subtract(CurrentTime, AppDataPtr->LastSquelchReturnedTime);
+
+        /* Calculate how many tokens to return in elapsed time */
+        if (DeltaTime.Seconds >= 1)
+        {
+            ReturnCount = CFE_PLATFORM_EVS_APP_EVENTS_PER_SEC;
+        }
+        else
+        {
+            ReturnCount = DeltaTime.Subseconds / (0xFFFFFFFF / CFE_PLATFORM_EVS_APP_EVENTS_PER_SEC);
+        }
+
+        /* Update last squelch returned time if we returned anything */
+        if (ReturnCount > 0)
+        {
+            AppDataPtr->LastSquelchReturnedTime = CurrentTime;
+        }
+
+        /* Return squelch tokens. */
+        if (ReturnCount > AppDataPtr->SquelchTokens)
+        {
+            AppDataPtr->SquelchTokens = 0;
+        }
+        else
+        {
+            AppDataPtr->SquelchTokens -= ReturnCount;
+        }
+
+
+        if (AppDataPtr->SquelchTokens > CFE_PLATFORM_EVS_MAX_APP_EVENT_BURST)
+        {
+            /*
+	     * Tokens equal to squelch rate, just squelch and increment squelch
+	     * counter, but not token counter
+	     */
+            NotSquelched = false;
+            AppDataPtr->SquelchedCount++;
+        }
+        else if (AppDataPtr->SquelchTokens == CFE_PLATFORM_EVS_MAX_APP_EVENT_BURST)
+        {
+            AppDataPtr->SquelchedCount++;
+            AppDataPtr->SquelchTokens++;
+            NotSquelched = false;
+
+            /* Just crossed threshold, send event message 
+	     * Note that this could cause spammage if the tokens reached the threshold
+	     * and stayed there if messages were sent constantly at slightly
+	     * greater than the suppress rate, but it still won't spam greater
+	     * than max events per second.
+	     */
+            CFE_ES_GetAppName(AppName, EVS_AppDataGetID(AppDataPtr), sizeof(AppName));
+            EVS_SendEvent(CFE_EVS_SQUELCHED_ERR_EID, CFE_EVS_EventType_ERROR,
+                          "Events squelched, AppName = %s, SquelchedCount = %u", AppName,
+                          (unsigned int)AppDataPtr->SquelchedCount);
+        }
+        else
+        {
+            AppDataPtr->SquelchTokens++;
+            /* NotSquelched already true */
+        }
+    }
+    return NotSquelched;
+}
+
```

Comment:
General comments on this whole thing, it still seems really complicated to me.  It also seems a bit paradoxical to send an event to say an app is sending too many events right now.

I understand the desire/intent to use a clock difference to restore credit, but I would argue the units are too coarse, aside from ensuring you use a monotonic clock, not a settable clock, that I already noted.  

Instead of crediting one token per event, make each event "cost" (for example) 1000 tokens.  Then give credit per millisecond instead of per second.

Make sending an event be a decrement, not an increment.  This makes it easy to check by checking the sign of your token credit.  If its negative, you cannot send.  Likewise on the increment side you just need to cap how high it can go (e.g. 1000 * CFE_PLATFORM_EVS_MAX_APP_EVENT_BURST).

I'd also consider make the event notification happen on the transition from negative to positive, rather than positive to negative.  This way, your event will be more useful, in that it will show the actual number of events squelched during the spewing period.  (I am guessing, if I'm reading the code correctly, that if an app starts spewing events the event that comes out will say "1" because it comes out only at the first squelch, even if it was the first of 1000 such issues ... by waiting till it actually stops spewing, you'll report the actual number that were surpressed during the spew).

Reply:
I'm kinda hesitant to have no event message for too many events, even though it seems paradoxical, since otherwise one would have to be actively looking for the event-squelched counter in the per-app telemetry (which isn't currently resettable for some reason) to know why events are mysteriously disappearing.

Emitting it when the token count goes from negative to positive as you mentioned below seems like it'd address this paradox, since you're emitting the event when there's "room" (although you'll never emit it if spamming doesn't stop- I suppose could emit another event message if the 8-bit squelch count is about to roll over, or some other threshold)

It's already giving credit at a rate of 1 per 1s/CFE_PLATFORM_EVS_APP_EVENTS_PER_SEC by dividing out the subseconds. Logic does get simpler if using OS_TimeGetTotalMilliseconds since that's one number to compare rather than whole seconds + subseconds. 

Otherwise I agree w/ everything else, and thanks for the feedback!

Reply:
Although I've also debated this a bit internally, I'm in favor of leaving the event message in as an indication an app is sending too many events.  It's the app events that are flooding and causing the problem, not a single additional event that is well behaved.  It also helps in the event "stream" (log or console or ground reporting of event messages) to see why the flooding events went away, becomes immediately identifiable operationally vs an assumption that the issue causing the flooding might have actually stopped.  If a system is really on the margins where +1 event will cause an issue, then I'd suggest just reducing the flooding limit by 1 vs dropping this helpful additional event.